fix bug causing forall loop outside of GPUs to execute serially.

stonea opened the pull request:
This PR is for: https://github.com/Cray/chapel-private/issues/2412

Specifically I add new variable to locale models indicating whether they should partition iterations across sublocales. Have all subocales other than flat and GPU
set this to true.

To code reviewers:

* Let me know if you feel the name `localeModelPartitionsIterationsOnSublocales` is appropriate. It's wordy but I feel it's not something that frequently used and the self documenting nature outweighs the verbosity.
* Let me know if you think the compilerAssert I have in ChapelLocale is in the right place. It's there to serve as a sanity check for anyone who creates a new locale model.
* The changes in cg-expr.cpp will introduce new errors with the NUMA local model in paratests (not with the flat locale model however). We already have several tests that fail with NUMA, so for this PR I'm accepting this as acceptable fallout, but if we expect that the NUMA model is anything other than experimental it may be a good idea to fork a new Github issue. Let me know what you think.
* I do have a test I used locally to test this (I'll copy it into the comments). But I'm not sure if this is appropriate to add to our test suite as I don't think we want to lock down how many NUMA domains we have in our testing hardware.

stonea commented:
Not sure where to stick this so I'm adding it as a comment here. But this is what I used to test these changes. I'm not sure how to add this to our testing system as I don't think we want to lock down the number of sublocales, but let me know if you disagree.

<details>

```chpl
proc getSubloc() {
  if localeModelHasSublocales {
    return chpl_getSubloc();
  }
  return 0;
}

writeln("NUM LOCALES: ", numLocales-1);
writeln("CHILD COUNT: ", here.getChildCount());
writeln("-------------");

{
  writeln("Range iteration:");
  var r1 = 1..4;
  forall i in r1 {
    writeln("  i = ", i, "\t", here, "\t", getSubloc());
  }
}

{
  writeln("Default rectangular domain zippered iteration:");
  var D : domain(2) = {1..2, 1..2};
  forall i in zip(D, D) {
    // TODO: For some reason if we use here (the constant) rather than "here"
    // the string we get the following error: use of 'here' before encountering
    // its definition, type unknown.  Looks like this is a bug that should be
    // fixed.  Once this is resolved remove the double quotes.
    // FIXME: See https://github.com/chapel-lang/chapel/issues/18778
    writeln("  i = ", i, "\t", "here", "\t", getSubloc());
  }
}

{
  writeln("Array view slice iteration:");
  const A = [1, 2, 3, 4, 5, 6];
  const D = {2..5};
  forall i in A[D] {
    writeln("  i = ", i, "\t", here, "\t", getSubloc());
  }
}

{
  writeln("Array view reindex iteration:");
  const A = [1, 2];
  const D = {4..5};
  forall i in A.reindex(D) {
    writeln("  i = ", i, "\t", here, "\t", getSubloc());
  }
}

{
  writeln("Array view rank change iteration:");
  var A : [1..2, 1..2] int;
  A[1, 1] = 1; A[1, 2] = 2; A[2, 1] = 3; A[2, 2] = 4;
  const B = A[1, ..];
  forall i in B {
    writeln("  i = ", i, "\t", here, "\t", getSubloc());
  }
}
```

Output on chapcs11:

```
NUM LOCALES: 0
CHILD COUNT: 2
-------------
Range iteration:
  i = 1	LOCALE0.ND0	0
  i = 4	LOCALE0.ND1	1
  i = 2	LOCALE0.ND0	0
  i = 3	LOCALE0.ND1	1
Default rectangular domain zippered iteration:
  i = ((2, 1), (2, 1))	here	1
  i = ((1, 1), (1, 1))	here	0
  i = ((1, 2), (1, 2))	here	0
  i = ((2, 2), (2, 2))	here	1
Array view slice iteration:
  i = 6	LOCALE0.ND1	1
  i = 5	LOCALE0.ND1	1
  i = 4	LOCALE0.ND0	0
  i = 3	LOCALE0.ND0	0
Array view reindex iteration:
  i = 1	LOCALE0.ND0	0
  i = 2	LOCALE0.ND1	1
Array view rank change iteration:
  i = 1	LOCALE0.ND0	0
  i = 2	LOCALE0.ND1	1
```
</details>

e-kayrakli commented:
> Let me know if you feel the name localeModelPartitionsIterationsOnSublocales is appropriate. It's wordy but I feel it's not something that frequently used and the self documenting nature outweighs the verbosity.

I think similarly. I don't love the name. I don't have a better alternative. I don't think it matters too much.

I would like to get a test for this. The code you have there makes sense only in NUMA (and APU, but that matters even less, I think), right? I see some alternatives:

- Ideal: Paramterize the correctness based on child count. You can make arrays (or dimensions) always numChildren*2, that should give you a guarantee about the sum of subloc id's you'll get from your iterations. Then you can change your foralls into reductions and check the result of the reduction. (I don't think it'd worth the effort)
- Assuming that this only makes sense for NUMA locale model, you can halt if the child count !=2 with a proper error message indicating that this test was designed for 2-numa-subdomain systems. At least we'll have an assurance that nobody would be puzzled if this test was run elsewhere.

As for the order of the good file: I think the order of lines is indeterministic. But you can have a prediff `sort` the output. I think lines themselves are deterministic. So it should work fine.

e-kayrakli reviewed (approved):
I'm good with changes here. I think this can go in once we have the correctness test in.

stonea commented:
Thanks Engin!  I've added a test with a `skipif` file so it only runs with the NUMA locale model and I've added code to error out if the number of sublocales isn't two. Potentially I could have the `.skipif` skip if number of sublocales isn't 2 but I'm afraid we may end updating our hardware and having the test never run.

Anyway, this should be fine for now (I believe we only run the with `CHPL_LOCALE_MODEL=numa` on `chapcs`), but potentially we'll have to update it once we get new testing hardware.  This is admittedly technical debt but I think leaving a test out would be a greater evil: I suspect a lack of NUMA testing is why we didn't notice the `coforall` related error in `cg-expr.cpp` earlier.

I think I'll hold off submitting the PR until after 1.25.1 is good.

stonea commented:
The title of this PR (and the merge message) indicate that I: "fix(ed a) bug causing forall loop outside of GPUs to execute serially"

Which is true for the NUMA locale model. Given case 1 from this comment I'm thinking this may be generally true as well: https://github.com/Cray/chapel-private/issues/2412#issuecomment-969530725

But haven't foralls in normal cases have generated parallelsim for years?  So maybe this was a temporary bug that existed or only applies to the NUMA locale model.

Either way when it comes time for release notes \ updating the changes.md file I think we should mark this as either being NUMA specific or verify if this bug really is as general as my title makes it out to be.

e-kayrakli commented:
> Edit: I'm thinking now maybe what case 1 (from that comment) was indicating that the forall loop would execute sequentially if outside of an on statement and when CHPL_LOCALE_MODEL=gpu.

Right -- If you have `CHPL_LOCALE_MODEL=gpu`, you'd still want your non-gpu `forall`s to execute in parallel on the host. This PR fixed that in my view.
